design/BufferOperations.md

```diff
@@ -0,0 +1,147 @@
+    WebGPUMappedMemory mapReadAsync(u32 offset, u32 size);
+    WebGPUMappedMemory mapWriteAsync(u32 offset, u32 size);
+};
+```
+
+Upon success the buffer is put in the mapped state and the call returns a `WebGPUMappedMemory` memory range.
+The return value starts in the pending state and represents the buffer range `[offset, offset + size)` for reading (for `mapReadAsync`) or writing (for `mapWriteAsync`).
+Upon failure a Javascript exception is raised (**TODO**: is this painting us in a corner for multiprocess? We could have an extra "error" state for WebGPUMappedMemory like other WebGPU objects).
+
+The following must be true or the mapping call on `buffer` fails:
+
+ - `buffer` must have been created with the `MAP_READ` usage flag for `mapReadAsync` and the `MAP_WRITE` flag for `mapWriteAsync`
+ - `offset + size` must be less or equal to the size of `buffer`.
+ - `buffer` must not be in the unmapped state.
+ - the range defined by `[offset, offset + size)` must not interesect the same range for `WebGPUMappedMemory` objects on the same buffer that aren't in the invalidated state.
+
+A buffer can be unmapped with:
+
+```webidl
+partial interface WebGPUBuffer {
+    void unmap();
+};
+```
+
+Upon succes the buffer is put in the unmapeped state, and all `WebGPUMappedMemory` objects created from the buffer are put in the invalidated state.
+The buffer is also put in the unmapped state.
+
+The following must be true or the unmapping call on `buffer` fails:
+
+ - `buffer` must have been created with the `MAP_READ` or the `MAP_WRITE` usage flags.
+ - `buffer` must not be in the destroyed state (this means it is ok to call `unmap` on an unmapped buffer).
+
+### `WebGPUMappedMemory` objects
+
+A `WebGPUMappedMemory` is essentially a wrapper around an `ArrayBuffer` that represents the content and controls its visibility.
+
+```webidl
+    readonly attribute ArrayBuffer? pointer;
+};
+```
+
+Because GPU operations can still be using the content of the buffer, `WebGPUMappedMemory` content is not visible immediately.
+There are two paths to know when the content of the buffer is visible: a synchronous path using `isPending` and an asynchronous path using `promise`.
+
+```webidl
+partial interface WebGPUMappedMemory {
+    readonly attribute bool isPending;
+    readonly attribute Promise<void> promise;
+}
+```
+
+`WebGPUMappedMemory` objects have an internal state machine that has three states:
+
+ - **Pending**: the content of the buffer is not yet visible.
+   In that state `pointer` is `null`, `isPending` is `true` and `promise` is unresolved.
+ - **Available**: the content of the buffer is visible.
+   In that state `pointer` is an `ArrayBuffer` representing the content of the buffer, `isPending` is `false` and `promise` is resolved.
+ - **Invalidated**: the content of the buffer is no longer visible.
+   In that state `pointer` is `null`, `isPending` is `false` and `promise` is rejected.
+
+State transitions are the following:
```

Comment:
You can only get a `WebGPUMappedMemory` by calling `WebGPUBuffer.map*`.
Since it's already mapped, using it in frame B would be an error since `unmap` has not been called, and it would still be `available`.